drivers/rtc/rtc_stm32.c

```diff
@@ -0,0 +1,322 @@
+{
+	struct device * const dev = (struct device *)userdata;
+	struct rtc_stm32_data *context = dev->driver_data;
+
+	HAL_RTC_AlarmIRQHandler(&context->hrtc);
+	rtc_data.cb_fn(dev);
+}
+#endif
+static const struct rtc_driver_api rtc_api = {
+		.enable = rtc_stm32_enable,
+		.disable = rtc_stm32_disable,
+		.read = rtc_stm32_read,
+		.set_config = rtc_stm32_set_config,
+		.set_alarm = rtc_stm32_set_alarm,
+		.get_pending_int = rtc_stm32_get_pending_int,
+};
+
+#ifdef CONFIG_DEVICE_POWER_MANAGEMENT
+
+#ifdef CONFIG_SYS_POWER_DEEP_SLEEP
+static int rtc_suspend_device(struct device *dev)
+{
+	__HAL_RCC_RTC_DISABLE();
+	rtc_stm32_set_power_state(dev, DEVICE_PM_SUSPEND_STATE);
+	return 0;
+}
+
+static int rtc_resume_device(struct device *dev)
+{
+	__HAL_RCC_RTC_ENABLE();
+	rtc_stm32_set_power_state(dev, DEVICE_PM_ACTIVE_STATE);
+
+	return 0;
+}
+#endif
+
+/*
+ * Implements the driver control management functionality
+ * the *context may include IN data or/and OUT data
+ */
+static int rtc_stm32_device_ctrl(struct device *dev, u32_t ctrl_command,
+				 void *context)
+{
+	if (ctrl_command == DEVICE_PM_SET_POWER_STATE) {
+#ifdef CONFIG_SYS_POWER_DEEP_SLEEP
+		if (*((u32_t *)context) == DEVICE_PM_SUSPEND_STATE) {
+			return rtc_suspend_device(dev);
+		} else if (*((u32_t *)context) == DEVICE_PM_ACTIVE_STATE) {
+			return rtc_resume_device(dev);
+		}
+#endif
+	} else if (ctrl_command == DEVICE_PM_GET_POWER_STATE) {
+		*((u32_t *)context) = rtc_stm32_get_power_state(dev);
+		return 0;
+	}
+
+	return 0;
+}
+#endif
+static int rtc_stm32_init(struct device *dev);
+DEVICE_DEFINE(rtc_stm32, CONFIG_RTC_0_NAME,
+	      &rtc_stm32_init, rtc_stm32_device_ctrl,
+	      &rtc_data, NULL, POST_KERNEL,
+	      CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
+	      &rtc_api);
+
+
+static int rtc_stm32_init(struct device *dev)
+{
+	RCC_OscInitTypeDef        RCC_OscInitStruct;
+	RCC_PeriphCLKInitTypeDef  PeriphClkInitStruct;
+
+	if (IS_ENABLED(CONFIG_RTC_STM32_API_REENTRANCY)) {
+		k_sem_init(RP_GET(dev), 0, UINT_MAX);
+		k_sem_give(RP_GET(dev));
+	}
+
+#ifdef CONFIG_SOC_SERIES_STM32F4X
+	stm32_exti_set_callback(/*STM32F4_IRQ_RTC_ALARM*/17,
+				rtc_stm32_isr, (void *)dev);
+#endif
+#ifdef CONFIG_SOC_SERIES_STM32L4X
+	IRQ_CONNECT(STM32L4_IRQ_RTC_ALARM, CONFIG_RTC_0_IRQ_PRI,
+		    rtc_stm32_isr, DEVICE_GET(rtc_stm32), 0);
+#endif
+	__HAL_RCC_PWR_CLK_ENABLE();
+	HAL_PWR_EnableBkUpAccess();
+
+
+	/* Configure LSI as RTC clock source */
+	RCC_OscInitStruct.OscillatorType = RCC_OSCILLATORTYPE_LSI;
+	RCC_OscInitStruct.PLL.PLLState = RCC_PLL_NONE;
+	RCC_OscInitStruct.LSIState = RCC_LSI_ON;
+	if (HAL_RCC_OscConfig(&RCC_OscInitStruct) != HAL_OK) {
+		return -EIO;
+	}
+
+	PeriphClkInitStruct.PeriphClockSelection = RCC_PERIPHCLK_RTC;
+	PeriphClkInitStruct.RTCClockSelection = RCC_RTCCLKSOURCE_LSI;
+	if (HAL_RCCEx_PeriphCLKConfig(&PeriphClkInitStruct) != HAL_OK) {
+		return -EIO;
+	}
+
+	__HAL_RCC_RTC_ENABLE();
+
+	rtc_data.hrtc.Instance = (RTC_TypeDef *) RTC_BASE;
```

Comment:
RTC_BASE info should be taken from device tree

Reply:
Done